### ISSUE TITLE
[Feature Request] Add support for Hidden Markov Models in torch.distributions

### ISSUE THREAD

**Issue by elSomewhere** [pytorch/pytorch#45133]
## 🚀 Feature
It would be useful to have a basic abstraction class for [Hidden Markov Models](https://en.wikipedia.org/wiki/Hidden_Markov_model) which builds on top of the distribution classes already implemented in the [pytorch distributions](https://pytorch.org/docs/stable/distributions.html#mixturesamefamily) package.

## Motivation
Hidden Markov Models (HMM) are a fundamental concept in many probabilistic applications. Custom implementations can already be found in existing Pytorch projects, the majority of which however do not take advantage of the modular [Distribution and Mixture classes](https://pytorch.org/docs/stable/distributions.html#distribution) in the distributions package.

The general concept of an HMM offers sufficient abstraction capabilities to justify a default implementation which builds upon these modules in the distributions package.

## Pitch
[In line with the decision to roughly follow the Tensorflow API](https://github.com/pytorch/pytorch/issues/18125#issuecomment-509921430) in terms of implementation of Distributions and derived families such as [Mixtures](https://pytorch.org/docs/stable/distributions.html#mixturesamefamily), I propose to port Tensorflow's [HiddenMarkovModel](https://www.tensorflow.org/probability/api_docs/python/tfp/distributions/HiddenMarkovModel) class to pytorch.

**Comment by fritzo** [comment]
Hi @elSomewhere, I agree HMMs are super useful. Here are some existing PyTorch implementations of HMMs:
- Pyro's [DiscreteHMM](http://docs.pyro.ai/en/stable/distributions.html#discretehmm) and [GaussianHmm](http://docs.pyro.ai/en/stable/distributions.html#gaussianhmm) classes.
- pytorch-struct's [HMM](http://nlp.seas.harvard.edu/pytorch-struct/model.html#Hidden-Markov-Model) class.

These PyTorch classes all define interfaces to cover specific classes of parallel-scan inference algorithms, which are 1-2 orders of magnitude faster than naive sequential inference. We could probably provide a tfp-like interface to Pyro's `DiscreteHMM` by using some pattern matching code. I'd be happy to provide pointers to anyone interested in implementing this. WDYT?

**Comment by elSomewhere** [comment]
Hi @fritzo 

Thanks for the Pyro HMM pointer - it actually appears to be reasonably close to the tensorflow setup, except the way transition and initialization distributions are defined (they are strictly categorical in tensorflow - it's nice to have all parameters inherit the same class, but in my view nothing game changing).

However, would it not still make sense to have a dedicated native HMM class in pytorch? I am new to the pytorch dev community, so glad to be proven wrong on any of these points. But I would reason as follows:
- Pyro embraces a fully probabilistic / bayesian approach. It may however be a bit overkill to introduce it as a dependency for those who just want a simple HMM class, which in my view is still a fairly basic building block for models.
- Given the abundance of HMM models out there being trained using "classic" approaches such as BaumWelch, having an easily accessible dedicated HMM class in pytorch that is explained and trained via gradient descent seems to be a nice core addition to the default framework and can highlight the different approaches available due to autograd.
- It would make sense then to refer to Pyro as soon as massive parallelism and / or fully bayesian parametrizations become necessary. The latter does require somewhat of a new mindset to model development, which then can be fully adopted once such a dependency is introduced (at least, this is how my workplace plans to handle a potential to transition to Pyro since there are some obvious advantages)

I would be glad to contribute to a native module. I've been working on a pytorch HMM implementation which follows the structure of torch.distributions and is close to the Tensorflow implementation. Basic functionality (including a short example of training all hmm parameters with backprop) is implemented and can be found [here](https://github.com/elSomewhere/New-PyTorch-modules/blob/master/HiddenMarkovModel.py).

Cleanup, boilerplate, some inherited methods, tests and closer ties to pytorch implementation standards may still be required, but I am glad to put some time in this and create a proper pull request.

**Comment by fritzo** [comment]
@elSomewhere note you can `import pyro.distributions as dist` and use `dist.DiscreteHMM` in pure PyTorch code without adopting any of Pyro's Bayesian methodology

**Comment by fritzo** [comment]
I think adding some sort of HMM distribution is a good idea, and serves as good motivation for some of the testing cleanup that @neeraj and I are planning to do. In particular because this distribution has unusually many inputs, it is helpful to test against a wide array of input shapes (e.g. Pyro's `DiscreteHMM` has [hundreds of tests](https://github.com/pyro-ppl/pyro/blob/7768099c444485d7404f6564b7b376ec6b219aa6/tests/distributions/test_hmm.py#L106)). This kind of testing is much clearer to express in pytest #11578, and will be much easier to do once we split `test_distributions.py` into multiple files.

I would also really like to implement a parallel-scan version in PyTorch rather than the naive sequential version, but I suppose this can wait for a follow-up PR. The parallel scan version parallelizes over the time dimension, and is helpful even if there is only a single time series with small transition matrices. In practice this is [easy to implement](https://github.com/pyro-ppl/pyro/blob/7768099c444485d7404f6564b7b376ec6b219aa6/pyro/distributions/hmm.py#L51) and makes a real difference in compute time for sequences as short as 100 time steps. In theory this even makes HMMs faster than nonassociative sequential models like RNNs, on long time series. For details on the interesting algorithms behind time-parallel inference, see
> Simo Sarkka, Angel F. Garcia-Fernandez (2019)
> "Temporal Parallelization of Bayesian Filters and Smoothers"
> https://arxiv.org/pdf/1905.13002.pdf
